### ISSUE TITLE
UP031 does not detect cases where percent operator is used on an f-string

### ISSUE THREAD

**Issue by richardxia** [astral-sh/ruff#13018]
This is admittedly a bit of a weird case, but I saw it in the wild and was surprised that Ruff wasn't flagging it as a lint error. With the `UP` ruleset enabled, Ruff isn't detecting the case where the `%` formatting operator is used on an f-string. Here's an example:

```python
def percent_formatting(x: str) -> str:
    return "%s" % x


def percent_formatting_and_f_string(x: str, y: str) -> str:
    return f"{x} %s" % y
```

Ruff will correctly flag the first function as having an issue, but the second one it silently ignores. I've created a playground link for this: https://play.ruff.rs/d0eec036-39bb-4db1-b55c-2dde7babaf7d

What I'd ideally expect is for Ruff to flag the second case as an issue and to suggest that the argument to the `%` be folded into the f-string, as `f"{x} {y}"`. Even if it can't suggest that specific fix, I think it would good to at least report that there is an error and that the author shouldn't be mixing formatting types like that.

**Comment by MichaReiser** [comment]
That makes sense. Thanks for reporting this

**Comment by dhruvmanila** [comment]
I don't think I would consider this as a bug because we need to expand the scope of the rule to cover f-strings as currently it only checks for string literals.

**Comment by MichaReiser** [comment]
I'm not very opinionated whether it's a bug or not. My take here is that, as a user, I would find it surprising that UP031 doesn't flag f-strings.

**Comment by Avasam** [comment]
Whether a different rule or an expansion of the existing one, I'm +1 on wanting this.

Regarding autofixes, I think the same as current `UP031` autofixes rules apply? Where Ruff needs to know the exact number of params to pass to the string (whether it's a tuple, a literal tuple, a single item, ...)

In OP's `percent_formatting_and_f_string` example, if  `x = "%s"` then it would already be failing.

**Comment by MeGaGiGaGon** [comment]
One additional wrinkle with this is that f-strings are evaluated before the `%` formatting, so you could have a situation where an f-string replacement is inserted to add percent formats, ie
```py
def example(insert, value):
    print(f"{insert}" % value)
example("%s", 1)
example("%s %s", (1, 2))
```
I'm not sure where you would run into this in normal code, but the fact it can happen makes it hard to say any f-string with a replacement isn't going to add/remove formats. This sounds to me like it's the same reason why `UP031` doesn't lint on complex expressions like concatenations, ie `("%s" + x) % y` https://play.ruff.rs/af9fdd40-679d-4844-b68a-21edd5562689
